File: test_fixtures/typescript/vulnerable.ts
// Zentinel ground truth — every line with a // TRIGGER comment should fire the named rule.
// Lines without TRIGGER should NOT fire any rule.
import { exec } from 'child_process';
import * as child_process from 'child_process';
import * as crypto from 'crypto';
import * as http from 'http';

// ── Code Injection ──────────────────────────────────────
function evalInput(code: string): void {
    eval(code);                                     // TRIGGER typescript.security.eval-usage
}

function funcConstructor(code: string): Function {
    return Function(code);                          // TRIGGER typescript.security.function-constructor
}

function delayedEval(code: string): void {
    setTimeout(code, 1000);                         // TRIGGER typescript.security.settimeout-string
}

function repeatedEval(code: string): void {
    setInterval(code, 1000);                        // TRIGGER typescript.security.setinterval-string
}

// ── Command Injection ───────────────────────────────────
function runCommand(cmd: string): void {
    exec(cmd);                                      // TRIGGER typescript.security.exec-usage
}

function cpExec(cmd: string): void {
    child_process.exec(cmd);                        // TRIGGER typescript.security.child-process-exec
}

function cpSpawn(cmd: string): void {
    child_process.spawn(cmd);                       // TRIGGER typescript.security.child-process-spawn
}

// ── XSS / DOM ───────────────────────────────────────────
function writeHtml(html: string): void {
    document.write(html);                           // TRIGGER typescript.security.innerhtml
}

// ── Cryptography ────────────────────────────────────────
function hashData(data: string): void {
    crypto.createHash("md5");                       // TRIGGER typescript.security.crypto-createhash-md5 + crypto-createhash-md5-precise
}

function hashSha1(data: string): void {
    crypto.createHash("sha1");                      // TRIGGER typescript.security.crypto-createhash-sha1 (+ crypto-createhash-md5)
}

// ── Tier 1: Template Literal Injection ──────────────────
function evalTemplate(code: string): void {
    eval(`run(${code})`);                           // TRIGGER typescript.security.eval-template (+ eval-usage)
}

// ── Hardcoded Secrets ───────────────────────────────────
const API_KEY: string = "sk-1234567890abcdef";      // TRIGGER typescript.security.hardcoded-secret
const DB_PASSWORD: string = "hunter2";              // TRIGGER typescript.security.hardcoded-secret

// ── Dangerous Patterns ──────────────────────────────────
function shutdown(): void {
    process.exit(1);                                // TRIGGER typescript.security.process-exit
}

File: test_fixtures/go/vulnerable.go
// Zentinel ground truth — every line with a // TRIGGER comment should fire the named rule.
// Lines without TRIGGER should NOT fire any rule.
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"database/sql"
	"fmt"
	"os"
	"os/exec"
)

// ── Command Injection ────────────────────────────────────
func runCommand(cmd string) {
	exec.Command(cmd).Run()                             // TRIGGER go.security.exec-command
}

// ── Dangerous Builtins ───────────────────────────────────
func unsafePrint(data interface{}) {
	fmt.Sprintf("%v", data)
}

// ── Cryptography ─────────────────────────────────────────
func hashMD5(data []byte) []byte {
	h := md5.Sum(data)                                  // TRIGGER go.security.crypto-md5
	return h[:]
}

func hashSHA1(data []byte) []byte {
	h := sha1.Sum(data)                                 // TRIGGER go.security.crypto-sha1
	return h[:]
}

// ── SQL Injection ────────────────────────────────────────
func queryUser(db *sql.DB, userID string) {
	query := fmt.Sprintf("SELECT * FROM users WHERE id = %s", userID)
	db.Query(query)
}

// ── Environment / Secrets ────────────────────────────────
var API_KEY = "sk-go-1234567890abcdef"              // TRIGGER go.security.hardcoded-secret
var DB_PASSWORD = "hunter2"                         // TRIGGER go.security.hardcoded-secret

// ── File Operations ──────────────────────────────────────
func readFile(path string) {
	os.ReadFile(path)
}

func main() {
	runCommand("ls -la")
	hashMD5([]byte("test"))
	hashSHA1([]byte("test"))
}
